### PR TITLE
EntityClass type annotations, and other minor stuff

### PR DESCRIPTION
**Pull request by csilvestrini**
Added some type annotations to the entity and storage proxy classes.
Deleted the Relation class because it was dead code (and then so was the Entity.data property).
Had to add a whole bunch of typecasts to unit tests, because things that were previously inferred to be `any` now have real types, and the compiler was correctly flagging issues with the way we were using them.

### CODE REVIEW

File: src/runtime/entity.ts [hunk 1]
```diff
@@ -12,16 +12,23 @@ import {Schema} from './schema.js';
 import {Symbols} from './symbols.js';
 import {Type} from './type.js';
 
+type EntityIdComponents = {
+  base: string,
+  component: () => number,
+};
+
+// tslint:disable-next-line: no-any
+export type EntityRawData = any;
```
Comment (shans) [inline src/runtime/entity.ts]: would {} work here instead of any?

File: src/runtime/handle.ts [hunk 2]
```diff
@@ -96,12 +102,12 @@ export abstract class Handle {
...
   }
 
-  _serialize(entity) {
+  _serialize(entity: EntityInterface) {
     assert(entity, 'can\'t serialize a null entity');
     if (!entity.isIdentified()) {
       entity.createIdentity(this._proxy.generateIDComponents());
     }
-    const id = entity[Symbols.identifier];
+    const id = entity.id;
```
Comment (shans) [inline src/runtime/handle.ts]: there was a notion at one stage that the id would be at entity[Symbols.identifier] so you could only extract it if you had access to Symbols.identifier. Not sure if we want to keep it or not; but switching to entity.id is a behaviour change.

File: src/runtime/entity.ts (hunk 1 above)
Comment (csilvestrini) [reply src/runtime/entity.ts]: Yes it does!

File: src/runtime/handle.ts (hunk 2 above)
Comment (csilvestrini) [reply src/runtime/handle.ts]: Oh is that how that works? Neat.

I was just experimenting with this, didn't mean to actually include this change. Reverted.

Comment (csilvestrini) [reply src/runtime/handle.ts]: Nope, I did it for a reason:

Type 'symbol' cannot be used as an index type.

See https://github.com/Microsoft/TypeScript/issues/1863

I've cast the symbol to any to get past this check. Any better ideas?

**Pull request merged**